task: jv-rick-and-morty. Executor: Oleksii Bashkirov.

OleksiiBashkirov opened the pull request:
feat: implement Rick and Morty character API
- Set up Spring Boot project with necessary dependencies.
- Added Character model and repository.
- Implemented CharacterClient to fetch and parse data from the Rick and Morty API.
- Created CharacterService to handle business logic.
- Developed CharacterController with endpoints for random character and search functionality.
- Configured Swagger for API documentation.
- Included necessary configurations and application properties.

OleksiiBashkirov committed:
feat: implement Rick and Morty character API

- Set up Spring Boot project with necessary dependencies.
- Added Character model and repository.
- Implemented CharacterClient to fetch and parse data from the Rick and Morty API.
- Created CharacterService to handle business logic.
- Developed CharacterController with endpoints for random character and search functionality.
- Configured Swagger for API documentation.
- Included necessary configurations and application properties.

okuzan commented on src/main/java/mate/academy/rickandmorty/controller/CharacterController.java:
```diff
@@ -0,0 +1,34 @@
+        return characterService.getRandomCharacter();
+    }
+
+    @GetMapping("/search")
```
```suggestion
    @GetMapping
```

okuzan commented on src/main/java/mate/academy/rickandmorty/service/CharacterClient.java:
```diff
@@ -0,0 +1,48 @@
+    private final ObjectMapper objectMapper;
+    private final HttpClient httpClient = HttpClient.newHttpClient();
+
+    public String fetchDataFromApi() {
+        try {
+            HttpRequest request = HttpRequest.newBuilder()
+                    .GET()
+                    .uri(new URI(url))
+                    .build();
+            HttpResponse<String> response = httpClient.send(
+                    request,
+                    HttpResponse.BodyHandlers.ofString()
+            );
+            return response.body();
+        } catch (IOException | InterruptedException | URISyntaxException e) {
+            throw new RuntimeException("Can't fetch data from API.", e);
+        }
+    }
```
you fetch only one page (20 characters)

OleksiiBashkirov committed:
Refactor Character API implementation to handle pagination

- Updated CharacterClient to fetch all characters from the API by handling pagination.
- Added PaginationInfo record to capture pagination details from API response.
- Updated CharacterServiceImpl to use fetchAllCharacters method and save all characters to the database.
- Ensured the application runs correctly and fetches all characters during initialization.

OleksiiBashkirov committed:
- Removed unnecessary "/search";

okuzan commented on src/main/java/mate/academy/rickandmorty/controller/CharacterController.java:
```diff
@@ -0,0 +1,34 @@
+        return characterService.getRandomCharacter();
+    }
+
+    @GetMapping()
```
```suggestion
    @GetMapping
```
about why we removed it - it's more RESTful, names should be resource oriented

okuzan commented on src/main/java/mate/academy/rickandmorty/controller/CharacterController.java:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("api/character")
```
```suggestion
@RequestMapping("characters")
```
API is a project wide prefix, better use server.servlet.context-path in properties to set it

OleksiiBashkirov replied:
Thanks for catching that! It looks like my application.properties already had server.servlet.context-path=/api, but I forgot to remove the /api prefix from @RequestMapping. I guess I got a bit too carried away with the double API  @action! I'll fix it right away.
